File: Trabalhos/2 - PontoFlutuante/PontoFlutuante_Java/src/br/ufba/mata49/trabalho/pontoflutuante/funcao/especial/SerieTaylor.java
package br.ufba.mata49.trabalho.pontoflutuante.funcao.especial;

import br.ufba.mata49.trabalho.pontoflutuante.funcao.base.FuncaoQuaternaria;
import br.ufba.mata49.trabalho.pontoflutuante.funcao.base.ComponenteSerie;

public class SerieTaylor implements FuncaoQuaternaria<Double, Double, Long, ComponenteSerie, Boolean> {

	private Potencia potencia;
	private Fatorial fatorial;
	
	public SerieTaylor() {
		potencia = new Potencia();
		fatorial = new Fatorial();
	}
	
	@Override
	public String getName() {
		return "Taylor";
	}
	
	public Double operar(Double x1, Long x2) {
		return operar(x1, x2, ComponenteSerie.TODOS, false);
	}

	@Override
	public Double operar(Double x1, Long x2, ComponenteSerie x3, Boolean alternada) {
		Double taylor = 0D;
		int iter = 0;
		boolean sinal = false;
		while(iter < x2) {
			if(x3 == ComponenteSerie.TODOS || (x3 == ComponenteSerie.PAR && iter % 2 == 0) || (x3 == ComponenteSerie.IMPAR && iter % 2 != 0))
				taylor += (((sinal=!sinal) || !alternada) ? 1 : -1) * (potencia.operar(x1, (long)iter)/(double)fatorial.operar((long)iter));
			iter++;
		}
		return taylor;
	}
	
}

File: Trabalhos/2 - PontoFlutuante/Pontoflutuante_cpp/funcao/especial/SerieTaylor.cpp
/*
 * SerieTaylor.cpp
 *
 *  Created on: 8 de dez de 2017
 *      Author: raffaello
 */

#include <iostream>
#include "SerieTaylor.h"

using namespace std;

std::string SerieTaylor::getName() {
	return "Taylor";
}

double SerieTaylor::operar(double x1, unsigned long int x2) {
	return operar(x1, x2, ComponenteSerie::TODOS, false);
}

double SerieTaylor::operar(double x1, unsigned long int x2, ComponenteSerie x3, bool alternada) {
	double taylor = 0;
	unsigned int iter = 0;
	bool sinal = false;
	while(iter < x2) {
		if((x3 == ComponenteSerie::TODOS) ||
				((x3 == ComponenteSerie::PAR) && (iter % 2 == 0)) ||
				((x3 == ComponenteSerie::IMPAR) && (iter % 2 != 0))) {
			sinal = !sinal;
			taylor += ((double)(sinal || !alternada) ? 1 : -1) * (potencia->operar(x1, iter)/fatorial->operarCpp(iter));
		}
		iter++;
	}
	return taylor;
}
